### PR TITLE
Remove double buffering in non-XDLOPS lowering process.

### PR DESCRIPTION
**Pull request by whchung**
- Remove double buffering in non-XDLOPS `gridwise_gemm` lowering path. This fixes the issue we crash when filter width/height are even numbers. The loop iteration is likely the culprit. Removing double buffering gets us easier main loop logic in `-miopen-lowering-step2`.
- Reorder e2e unit tests.
- Fix convolution crashes when filter width/height are even numbers.
- Add e2e unit tests to cover 2x2 and 4x4 filter sizes.

### CODE REVIEW

**Comment by whchung** [comment]
@yiqian1 I added additional e2e tests for filter width/height of 2 and 4.

**Comment by whchung** [comment]
Looking into test failures.

**Comment by whchung** [comment]
@yiqian1 This PR brings unexpected side effects to this layout config: YXCK/NHWC/NHWK. All related tests are failing. On the other hand, KCYX/NCHW/NKHW and KYXC/NHWC/NHWK configs are passing. I've temporarily disable YXCK/NHWC/NHWK configs as they will not be used for the June target.

**Comment by yiqian1** [comment]
Test-related changes look good to me. I'll let other people review gridwise_gemm changes.

File: mlir/test/mlir-miopen-driver/e2e/conv2d_harness_cpu_verifier_kyxc_nhwc_nhwk.mlir [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+    %10 = alloc() : memref<128x30x30x128xf32>
+    %11 = memref_cast %10 : memref<128x30x30x128xf32> to memref<?x?x?x?xf32>
+    call @mcpuMemset4DFloat(%11, %cst_0) : (memref<?x?x?x?xf32>, f32) -> ()
+
+    // launch cpu convolution
+    call @conv2d_host(%6, %8, %10) : (memref<128x3x3x8xf32>, memref<128x32x32x8xf32>, memref<128x30x30x128xf32>) -> ()
+
+    // verity results
+    call @verify_results(%10, %2) : (memref<128x30x30x128xf32>, memref<128x30x30x128xf32>) -> ()
```
Comment (jerryyin) [inline mlir/test/mlir-miopen-driver/e2e/conv2d_harness_cpu_verifier_kyxc_nhwc_nhwk.mlir:44]: Shall we use `-ph` instead of manually populating those?

**Comment by jerryyin** [comment]
Test related change looks good, abstain from gridwisegemm lowering.

File: mlir/test/mlir-miopen-driver/e2e/conv2d_harness_cpu_verifier_kyxc_nhwc_nhwk.mlir (hunk 1 above)
Comment (whchung) [reply mlir/test/mlir-miopen-driver/e2e/conv2d_harness_cpu_verifier_kyxc_nhwc_nhwk.mlir:44]: This test was created deliberately to skip `-ph`. In fact it was the source how `ph` was created.

**Comment by whchung** [comment]
@asroy ping. Please help review this PR. The idea behind it is to remove double buffering from `gridwise_gemm` lowering logic. This avoid the bug when filter width/height are even numbers, the bottom half of the double buffering loop still gets executed and the sliding window moves OOB.

**Comment by whchung** [comment]
Deep diving the code with @asroy now.

**Comment by whchung** [comment]
YXCK/NHWC/NHWK error is now fixed in #106 . Will revise this PR after #106 gets in.

**Comment by whchung** [comment]
jenkins: retest this please.

**Review by asroy — APPROVED** [review]

**Pull request merged**